introduce network id transaction body element

mkv-vcm opened the pull request:
Should we remove the networkId sent in the transaction init APDU?

mkv-vcm commented:
@janmazak so two questions:

1. element 15 is not mandatory in the tx body, but we add it all the time. Should the txHashBuilder code handle it as optional like now, or should it assume it's always there?
2. what are these "By always including the network id, we can get rid of the warning and the **related logic**" parts that we can remove? We still want _some_ checks on the network id, I don't really find much code to throw out.

janmazak commented on src/securityPolicy.c:
```diff
@@ -276,9 +276,11 @@ security_policy_t policyForSignTxInit(
 		ASSERT(false);
 	}
 
+	DENY_IF(networkId > MAXIMUM_NETWORK_ID);
```
Why don't we use `isValidNetworkId`?

mkv-vcm replied:
True.

janmazak commented on src/txHashBuilder.c:
```diff
@@ -1180,10 +1183,46 @@ static void txHashBuilder_assertCanLeaveMint(tx_hash_builder_t* builder)
+
+	txHashBuilder_assertCanLeaveMint(builder);
+
+	// add fee item into the main tx body map
```
not fee

mkv-vcm replied:
True.

janmazak commented on src/txHashBuilder.h:
```diff
@@ -15,6 +15,7 @@ enum {
 	TX_BODY_KEY_AUX_DATA = 7,
 	TX_BODY_KEY_VALIDITY_INTERVAL_START = 8,
 	TX_BODY_KEY_MINT = 9,
+        TX_BODY_KEY_NETWORK_ID = 15,
```
Indentation.

mkv-vcm replied:
Done.

janmazak commented on src/txHashBuilder.c:
```diff
@@ -1180,10 +1183,46 @@ static void txHashBuilder_assertCanLeaveMint(tx_hash_builder_t* builder)
+	case TX_HASH_BUILDER_IN_NETWORK_ID:
+		break;
+
+	case TX_HASH_BUILDER_IN_MINT:
+	case TX_HASH_BUILDER_IN_VALIDITY_INTERVAL_START:
+	case TX_HASH_BUILDER_IN_AUX_DATA:
+	case TX_HASH_BUILDER_IN_WITHDRAWALS:
+	case TX_HASH_BUILDER_IN_CERTIFICATES:
+	case TX_HASH_BUILDER_IN_TTL:
+	case TX_HASH_BUILDER_IN_FEE:
+		txHashBuilder_assertCanLeaveMint(builder);
+		ASSERT(!builder->includeValidityIntervalStart);
+		break;
```
I'm confused by this. Is it true it will disappear if we make network id mandatory?

mkv-vcm replied:
Well, it would look like this: ASSERT(builder->state == TX_HASH_BUILDER_IN_NETWORK_ID); since there is no other way to leave it. Not much of a code reduction, but still some bytes.

janmazak replied:
I am mostly interested in the simplicity :)
Having `ASSERT(!builder->includeValidityIntervalStart);` in a function named `txHashBuilder_assertCanLeaveNetworkId` is just confusing to me, and it seems simpler to just get rid of it (by making network id always included) instead of thinking about why it is here, and whether it should be.

mkv-vcm replied:
There are already examples for this in other mandatory elements, like Fee. I think some kind of assert is needed, so the Confirm state can check if anything is missing.

janmazak commented on src/txHashBuilder.h:
```diff
@@ -52,7 +53,8 @@ typedef enum {
 	TX_HASH_BUILDER_IN_MINT_ASSET_GROUP = 1011,
 	TX_HASH_BUILDER_IN_MINT_TOKEN = 1012,
-	TX_HASH_BUILDER_FINISHED = 1100,
+	TX_HASH_BUILDER_IN_NETWORK_ID = 1100,
```
This might need some adjustment when other tx body fields are added so that we don't have any clash. I'd leave the FINISHED value alone, though.

mkv-vcm replied:
Not sure how to make it certain it causes a merge conflict and demands attention then.

janmazak replied:
I'm not sure either, but it seems to make sense to build all the other Alonzo features on top of this branch, and then we don't need to think about the conflict.

mkv-vcm replied:
Yeah, just occurred to me while finishing another alonso feature. Will rebase and rework that after the standup.

janmazak reviewed (commented):
1. Making it mandatory in txhashbuilder will probably allow us to save some code in the txhashbuilder state machine, won't it? (see detailed comments)
2. I agree that it won't be much code, probably only a handful of lines around the warning we are removing (I've not checked in detail yet).

So far, this only includes changes in the hash builder? And signTx changes will be done within this PR?

mkv-vcm commented:
> 2. I agree that it won't be much code, probably only a handful of lines around the warning we are removing (I've not checked in detail yet).

I think I modified all the relevant parts, but I might have missed some.

> So far, this only includes changes in the hash builder? And signTx changes will be done within this PR?

As far as I see, there is no need for much modification in signTx. That one line I added in this PR so far is all we need, really. Since it's a mandatory element and not tied to a APDU.

janmazak commented on src/securityPolicy.c:
```diff
@@ -276,7 +276,7 @@ security_policy_t policyForSignTxInit(
 	}
 
-	DENY_IF(networkId > MAXIMUM_NETWORK_ID);
+	DENY_IF(!isValidNetworkId(networkId));
```
```suggestion
	DENY_UNLESS(isValidNetworkId(networkId));
```

mkv-vcm replied:
Done.

janmazak commented on src/securityPolicy.c:
```diff
@@ -276,7 +276,7 @@ security_policy_t policyForSignTxInit(
-	DENY_IF(networkId > MAXIMUM_NETWORK_ID);
+	DENY_IF(!isValidNetworkId(networkId));
 
 	WARN_IF(!is_tx_network_verifiable(numOutputs, numWithdrawals, txSigningMode));
```
```suggestion
	WARN_UNLESS(is_tx_network_verifiable(numOutputs, numWithdrawals, txSigningMode));
```

mkv-vcm replied:
Done.

janmazak commented on src/securityPolicy.c:
```diff
@@ -276,9 +276,11 @@ security_policy_t policyForSignTxInit(
+	DENY_UNLESS(isValidNetworkId(networkId));
 
-	WARN_IF(networkId != MAINNET_NETWORK_ID);
+	WARN_UNLESS(is_tx_network_verifiable(numOutputs, numWithdrawals, txSigningMode));
```
We aimed to get rid of `is_tx_network_verifiable` --- if tx body contains network id explicitly, we don't care whether it is included in ouputs or wihdrawals.

Also, the warning
```
			// technically, no withdrawals/pool reg. certificate as well, but the UI message would be too long
			ui_displayPaginatedText(
			        header,
			        "no outputs, cannot verify network id",
			        this_fn
			);

```
in signTx.c is no longer needed --- that's the "related logic" I mentioned before.

mkv-vcm replied:
Ah yes, since we always know what the networkId is, or at least code it into the tx. Ok, done!

janmazak commented on src/securityPolicy.c:
```diff
@@ -239,7 +218,6 @@ security_policy_t policyForSignTxInit(
         uint8_t networkId,
         uint32_t protocolMagic,
         uint16_t numInputs MARK_UNUSED,
-        uint16_t numOutputs,
```
I'd say we should either include all the numbers here (and mark them MARK_UNUSED if appropriate), or use just those that are relevant (I have a slight preference for this), i.e. we can remove numInputs from params.

mkv-vcm replied:
I absolutely on the side of removing unnecessary parameters.

janmazak commented on src/signTx.c:
```diff
@@ -338,20 +338,11 @@ static void signTx_handleInit_ui_runStep()
 		        "New script transaction" :
```
Not for this PR, but when renaming the singing mode SCRIPT -> MULTISIG, we want to change this string --- please note it somewhere so that we do not forget. If "New multisig transaction" is too long to fit on the display, I'd go with "Multisig transaction".

mkv-vcm replied:
Will try.

janmazak commented on src/txHashBuilder.c:
```diff
@@ -1180,10 +1183,46 @@ static void txHashBuilder_assertCanLeaveMint(tx_hash_builder_t* builder)
+	case TX_HASH_BUILDER_IN_NETWORK_ID:
+		break;
+
+	case TX_HASH_BUILDER_IN_MINT:
+	case TX_HASH_BUILDER_IN_VALIDITY_INTERVAL_START:
+	case TX_HASH_BUILDER_IN_AUX_DATA:
+	case TX_HASH_BUILDER_IN_WITHDRAWALS:
+	case TX_HASH_BUILDER_IN_CERTIFICATES:
+	case TX_HASH_BUILDER_IN_TTL:
+	case TX_HASH_BUILDER_IN_FEE:
+		txHashBuilder_assertCanLeaveMint(builder);
+		ASSERT(!builder->includeValidityIntervalStart);
+		break;
+
```
```suggestion
```

mkv-vcm replied:
True.

janmazak reviewed (approved):
This should be squashed into a single commit and made part of the Alonzo phase 1 branch & PR.
